src/libAtomVM/bitstring.c

```diff
@@ -119,10 +140,73 @@ bool bitstring_utf8_encode(avm_int_t c, uint8_t *buf, size_t *out_size)
     return true;
 }
 
+bool bitstring_utf8_decode(int32_t *c, uint8_t *buf, size_t len, size_t *out_size)
+{
+    int32_t v = 0;
+    if (len == 0) {
+        return false;
+    } else if (len >= 4 && (*buf & 0xF8) == 0xF0 && ((*(buf + 1) & 0xC0) == 0x80) && ((*(buf + 2) & 0xC0) == 0x80) && ((*(buf + 3) & 0xC0) == 0x80)) {
+        v |= (*buf & 0x07) << 18;
+        v |= (*(buf + 1) & 0x3F) << 12;
+        v |= (*(buf + 2) & 0x3F) << 6;
+        v |= (*(buf + 3) & 0x3F);
+        if (invalid_codepoint(v)) {
+            return false;
+        }
+        // overlong encoding
+        if (v <= 0x10000) {
+            return false;
+        }
+        *c = v;
+        *out_size = 4;
+        return true;
+    } else if (len >= 3 && (*buf & 0xF0) == 0xE0 && ((*(buf + 1) & 0xC0) == 0x80) && ((*(buf + 2) & 0xC0) == 0x80)) {
+        v |= (*buf & 0x0F) << 12;
+        v |= (*(buf + 1) & 0x3F) << 6;
+        v |= (*(buf + 2) & 0x3F);
+        if (invalid_codepoint(v)) {
+            return false;
+        }
+        // overlong encoding
+        if (v <= 0x800) {
+            return false;
+        }
+        *c = v;
+        *out_size = 3;
+        return true;
+    } else if (len >= 2 && (*buf & 0xE0) == 0xC0 && ((*(buf + 1) & 0xC0) == 0x80)) {
+        v |= (*buf & 0x1F) << 6;
+        v |= (*(buf + 1) & 0x3F);
+        if (invalid_codepoint(v)) {
```

Comment:
This test is unnecessary.

Reply:
Yeah that makes sense, I guess since:
```
// | U+0080   | 110xxxxx | 10xxxxxx |          |          |
```

So we only have 12 bits to play with.